odoo/service/server.py

```diff
@@ -399,20 +399,41 @@ def process_limit(self):
             self.limit_reached_time = None
 
     def cron_thread(self, number):
+        # Steve Reich timing style with thundering herd mitigation.
+        #
+        # On startup, all workers bind on a notification channel in
+        # postgres so they can be woken up at will. At worst they wake
+        # up every SLEEP_INTERVAL with a jitter. The jitter creates a
+        # chorus effect that helps distribute on the timeline the moment
+        # when individual worker wake up.
+        #
+        # On NOTIFY, all workers are awaken at the same time, sleeping
+        # just a bit prevents they all poll the database at the exact
+        # same time. This is known as the thundering herd effect.
+
         from odoo.addons.base.models.ir_cron import ir_cron
-        while True:
-            time.sleep(SLEEP_INTERVAL + number)     # Steve Reich timing style
-            registries = odoo.modules.registry.Registry.registries
-            _logger.debug('cron%d polling for jobs', number)
-            for db_name, registry in registries.d.items():
-                if registry.ready:
-                    thread = threading.currentThread()
-                    thread.start_time = time.time()
-                    try:
-                        ir_cron._acquire_job(db_name)
-                    except Exception:
-                        _logger.warning('cron%d encountered an Exception:', number, exc_info=True)
-                    thread.start_time = None
+        conn = odoo.sql_db.db_connect('postgres')
+        with conn.cursor() as cr:
+            pg_conn = cr._cnx
+            cr.execute("LISTEN cron_trigger")
+            cr.commit()
+
+            while True:
+                select.select([pg_conn], [], [], SLEEP_INTERVAL + number)
+                time.sleep(number / 100)
+                pg_conn.poll()
```

Comment:
You could actually use the return value of `select` to figure out why you have been woken up.
```python
rlist, wlist, xlist = select.select([pg_conn], [], [], SLEEP_INTERVAL + number)
if rlist:
    # woken up by pg_conn -> process jobs of the given databases first
else:
    # woken up by timeout -> process jobs of all databases
```
You may also keep a "last processed" timestamp for each database, to avoid starvation in case you have been woken up by triggers several times in a row.

Reply:
We decided to always process all databases for now, I could have the `if rlist:` branch to avoid polling the db when there is nothing thought but there is no such thing as "process the given databases first" we don't even send them anymore.